core/common/src/Utf8.kt

```diff
@@ -196,24 +205,31 @@ public fun Source.readString(byteCount: Long): String {
 }
 
 /**
- * Removes and returns a single UTF-8 code point, reading between 1 and 4 bytes as necessary.
+ * Decodes a single code point value from UTF-8 code units, reading between 1 and 4 bytes as necessary.
  *
  * If this source is exhausted before a complete code point can be read, this throws an
  * [EOFException] and consumes no input.
  *
- * If this source doesn't start with a properly-encoded UTF-8 code point, this method will remove
- * 1 or more non-UTF-8 bytes and return the replacement character (`U+fffd`). This covers encoding
- * problems (the input is not properly-encoded UTF-8), characters out of range (beyond the
- * `0x10ffff` limit of Unicode), code points for UTF-16 surrogates (`U+d800`..`U+dfff`) and overlong
- * encodings (such as `0xc080` for the NUL character in modified UTF-8).
+ * If this source starts with an ill-formed UTF-8 code units sequence, this method will remove
+ * 1 or more non-UTF-8 bytes and return the replacement character (`U+fffd`).
+ *
+ * The replacement character (`U+fffd`) will be also returned if the source starts with a well-formed
+ * code units sequences, but a decoded value does not pass further validation, such as
+ * the value is of range (beyond the `0x10ffff` limit of Unicode), maps to UTF-16 surrogates (`U+d800`..`U+dfff`),
```

Comment:
```suggestion
 * the value is out of range (beyond the `0x10ffff` limit of Unicode), maps to UTF-16 surrogates (`U+d800`..`U+dfff`),
```

It's better now, but maybe my Unicode knowledge still fails me: why `a decoded value does not pass further validation` if it `maps to UTF-16 surrogates ('U+d800'..'U+dfff')`?

Reply:
> why a decoded value does not pass further validation if it maps to UTF-16 surrogates ('U+d800'..'U+dfff')

3. Conformance / § 3.9 Unicode Encoding Forms / D92
```
Because surrogate code points are not Unicode scalar values, 
any UTF-8 byte sequence that would otherwise map to code points U+D800..U+DFFF is ill-formed.
```

https://www.unicode.org/versions/Unicode15.1.0/ch03.pdf

Reply:
Ah, I guess the meaning is that 'byte sequence should form a whole code point, not a half of it' and 'valid utf-16 is not always a valid utf-8'.

Reply:
> byte sequence should form a whole code point, not a half of it

That's a funny part:
- Unicode code point is an integer in [0, 0x1FFFF] range (D10, D9 from 3.4 Characters and Encoding)
- Unicode scalar values are all code points except surrogates (D76 from 3.9 Unicode Encoding Forms)
- Surrogates should not be interpreted as abstract characters and are subject to restrictions in interchanges as per 2.4 Code Points and Characters.

But the abovementioned D92 simply restricts encoding surrogates using UTF-8, despite surrogates being (almost) regular Unicode code points.